paddle/framework/prune.cc

```diff
@@ -102,6 +103,32 @@ void prune_impl(const proto::ProgramDesc& input, proto::ProgramDesc* output,
       *op_field->Add() = input.blocks(block_id).ops(i);
     }
   }
+
+  // remove the VarDescs in BlockDesc that are not referenced in
+  // the pruned OpDescs
+  std::unordered_map<std::string, proto::VarDesc> var_map;
+  auto* var_field = output->mutable_blocks(block_id)->mutable_vars();
+  for (const auto& var : *var_field) {
+    var_map[var.name()] = var;
+  }
```

Comment:
Maybe, there is no need for an extra map. We can directly use [var map](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/block_desc.h#L106) of input ProgramDesc.  We can use `input.FindVar(name)->Proto()` to get the `proto::VarDesc`.

Reply:
There are two different classes one is framework::proto::xxxDesc, one is framework::xxxDesc.  The latter is basically a wrapper about the former. 

For this prune function, the input is proto::ProgramDesc, so we only have limited functionality provided by the protobuf library itself. After you compile the framework.proto, you will get framework.pb.h + cc, which lists all the available functions for ProgramDesc and etc. 

So I don't think we can use input.FindVar()

Reply:
Later, when we are going to implement out transpiler function for inference optimization on the c++ side after loading a inference desc from file or buffer, we can use your suggestion, as in that case, we can take a framework::ProgramDesc as input. 

@Xreki what do you think?

Reply:
> For this prune function, the input is `proto::ProgramDesc`, so we only have limited functionality provided by the protobuf library itself.

You are right. Sorry for didn't notice that.